src/helpers.rs

```diff
@@ -471,6 +471,15 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     }
 }
 
+/// Check that the number of args is what we expect.
+pub fn check_args<'a, 'tcx, const N: usize>(args: &'a [OpTy<'tcx, Tag>]) -> InterpResult<'tcx, &'a [OpTy<'tcx, Tag>; N]>
```

Comment:
I think this function should check more than just arity (e.g. every scalar that goes through here should at least pass`this.read_scalar(ptr)?.not_undef()?`). But probably every argument should be validated to be the correct type as well when possible. Some shims that ignore a flag or mask drop it on the floor, some at least validate that it's the correct type (and then ignore it). That should all probably be a separate issue though.

Reply:
Almost all shims use their argument, so I am not sure if it's worth repeating that information by forcing them to give an explicit signature here or so.

Note that not all arguments have to be scalars, so even `read_scalar` is going too far.

IMO we can leave it to each shim to validate the size of its arguments, to whatever extend is appropriate. Anyway UB detection for shims (this is basically library UB, not language UB) is on a best-effort basis.